# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

In [9]:
#whiteboard problem week3day1: find the average of a list, round down to nearest integer
youngGrades = [5,6,6,7,8,8,9,9,12]
def findAvg(grades):
    average = sum(grades)/ len(grades)
    return int(average)
print(findAvg(youngGrades))


7


In [18]:
#week2 project
#if you don't have a return statement, means return none
#class lets you make many instances of that object
class ParkingGarage():
    def __init__(self):
        self.tickets = []
        self.parkingSpaces = [0,1,2,3,4]
        self.currentTicket = {} #value False = not paid for, True = paid for
        return
   
    def takeTicket(self):
        try:
            spot = int(input(f'what spot do you want? Currently available: {self.parkingSpaces}\n'))
            self.parkingSpaces.remove(spot)
            self.tickets.append(spot)
            self.currentTicket[spot]= False      
        except:
            print('Invalid inpupt! Please enter a number')
        
    def payForParking(self):
        try:
            spot = int(input(f'Which spot do you want to pay for? \n Current Tickets: {self.tickets}'))
            if spot not in self.tickets:
                print('That ticket does not exist/does not need to be paid for')
            if self.currentTicket[spot]== True:
                print('That ticket is already paid for! You can leave.')
            else:
                self.currentTicket[spot] = True
                print(f'You have paid for ticket {spot}. Thank you, you are free to go.')
                         
        except:
            print('Invalid input! Please enter a number')
    
    def leaveGarage(self):
        try:
            spot = int(input(f'Which spot are your leaving?'))
            if spot not in self.tickets:
                print('That ticket does not exist, please enter the correct number')
            if self.currentTicket[spot] == False:
                print('That spot has not been paid for.. please pay then try again')
            else: 
                self.tickets.remove(spot)
                self.parkingSpaces.append(spot)
                del self.currentTicket[spot]
                self.parkingSpaces.sort()
                print('Thank you for using our parkign garage! Hope to see you soon.')
        except:
            print('Invalid input, please enter a number')
   
    def showAvailable(self):
        print('available spaces: ',self.parkingSpaces)
        print('tickets used: ',self.tickets)
        print('tickets status: ',self.currentTicket)
    
    def runSimulation(self):
        while True:
            answer = input('What do you want to do? (ticket/pay/leave/show/exit)')
            if answer.lower() == 'ticket':
                self.takeTicket()
            elif answer.lower()== 'pay':
                self.payForParking()
            elif answer.lower()== 'leave':
                self.leaveGarage()
            elif answer.lower()== 'show':
                self.showAvailable()
            elif answer.lower ()== 'exit':
                break
            else:
                print('Invalid response. Please type: ticket/pay/leave/show/exit')
        

garage1 = ParkingGarage()
garage1.runSimulation()

#these two lines are the only actual lines of code being run, can rerun simluations on garage2 w/ different capacity etc

What do you want to do? (ticket/pay/leave/show/exit)exit


#### Syntax

In [22]:
def addTwo(num):
    return num + 2
print(addTwo(4))

#Lambda function syntax
print((lambda num: num+2)(10))

#usually you define a function so you can use it again

6
12


#### Saving to a Variable

In [23]:
f_test = lambda x: x+2
f_test(10)

12

#### Multiple Inputs

In [26]:
# multiple input parameters not savings to variable
print((lambda x,y,z: x+y+z)(2,10,12))

# save to variable with multiple inputs
x_test = lambda x,y,z: z-y-x
print(x_test(1,6,19))


24
12


#### Passing a Lambda into a Function

In [32]:
def multiply(f, num):
    '''
    f expects a lambda function
    num expects a number 
    '''
    return f(num)
multiply(lambda x:x*x, 4)
#lamda saved to variable f and the x saved into the num


16

#### Returning a Lambda from a Function

In [37]:
# regular defiend function
def multiplyTest(num):
    return num * 4


def returnFunc():
    test = 4
    def multiply2(num):
        return num * 2
    return multiply2

fReturn = returnFunc() #saying fReturn = multiply2(num)
print(fReturn(4))

#the fReturn calls the multiply2 funciton

def returnLambda(b,c):
    return lambda x,y: x*y+b+c        #x,y are the input parameters

lReturn = returnLambda(4,5)    #this returns another function w/ 4,5 plugged into b,c
print(lReturn(2,3))               #2, 3 are being plugged into x and y
#lambda functions are helpful for call back functions 

8
15


#### If Statements within Lambdas

In [46]:
lCondition = lambda num: num*2 if num>10 else num +2
print(fCondition(12))

#variable_name = x: what to do to x, if statement

def fCondition(num):
    if num >10:
        return num *2
    else:
        return num +2
print(fCondition(10))



24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [47]:
f = lambda x: x**3
print(f(2))

print('think I can do it another way- function within a function type beat')



8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [141]:
#map(func, iter)
def square(x):
    if x <5:
        return x**2
    else:
        return x
output = list(map(square, [1,2,3,4,5,6,7,8]))
print(output)

print('==================================')
def regSquare(list1):
    output2 = []
    for num in list1:
        output2.append(num**2)
    return output2
print(regSquare([1,2,3,4,5]))




[1, 4, 9, 16, 5, 6, 7, 8]
[1, 4, 9, 16, 25]


#### Using Lambda's with Map

In [65]:
list(map(lambda x: x**2 if x<5 else x, [4,5,6,7,8]))

[16, 5, 6, 7, 8]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [69]:
list1 = [1,2,3,4,5,6]
list(map(lambda x: x*2-1,list1 ))

[1, 3, 5, 7, 9, 11]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [75]:
#map(func, iter): map adds the return value of func() to our new list
#filter(func, iter): adds teh same value from iter, based on wheter the return value of func() is true or not
def myFilter(num):
    if num%2==0:
        return False   #filtering out the even numbers 
    return True   #if not false, returns True

list(filter(myFilter, [1,2,3,4,5,6,7,8,9]))
#the variables in the function that return True are put into the list

# can't do myFilter([1,2,3,4,5]) bc you can't % a list unless you do a loop; filter loops for us

[1, 3, 5, 7, 9]

#### Using Lambda's with Filter()

In [76]:
list(filter(lambda x: False if x%2==0 else True, [1,2,3,4,5,6,7,8,9]))

[1, 3, 5, 7, 9]

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [89]:
import statistics

data = [1,2,3,4,5,6,7,8,9,10]
print(list(filter(lambda x: False if x < statistics.mean(data) else True, data)))

print('shoha way=============================')
answer = list(filter(lambda num: True if num>statistics.mean([1,2,3,4,5,6,7,8,9,10]) else False, [1,2,3,4,5,6,7,8,9,10]))
print(answer)


[6, 7, 8, 9, 10]
shoha way=============================
[6, 7, 8, 9, 10]


In [91]:
if 'yes':
    print('here')
#0 equates to False, any other number equates True
#anything that is not empty is inherently True

here


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [98]:
#used in javascript(week 6)
from functools import reduce
#reduce(func, iter)

list1 = [2,4,7,8,10]
def addNums(num1,num2):
    print('num1',num1)
    print('num2', num2)
    return num1+ num2

result = reduce(addNums, list1)  #reduces an iterable to a single item
print(result)                    


#map applies every iteraction to your function; 
#reduce: iterates through your list, applies your function and returns the aggregated value
#num1 =2, num2=4 so num1 = 6 then num2=7 which returns 13
#because it has been moved to 'functools' library, not as important 

num1 2
num2 4
num1 6
num2 7
num1 13
num2 8
num1 21
num2 10
31


#### Using Lambda's with Reduce()

In [99]:
reduce(lambda x,y: x*y, [1,2,3,4,5])

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [101]:
reduce(lambda x,y: x*y if x%2==0 else x//y, [6,2,3,4,5])

720

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [109]:
def addNums(num):
    #set a base case for a recursive function
    if num <= 1:
        print('addNums(1)=1')
        return num
    print(f'addNums({num})= {num} + addNums({num-1})')
    return num + addNums(num-1)     
    
print(addNums(6))
# does 6+5+4+3+2+1= 21, when we reach num <=1 then we return

addNums(6)= 6 + addNums(5)
addNums(5)= 5 + addNums(4)
addNums(4)= 4 + addNums(3)
addNums(3)= 3 + addNums(2)
addNums(2)= 2 + addNums(1)
addNums(1)=1
21


In [112]:
def infinite(num):
    if num ==1:
        return 'done'
    print(f'{num}, looping again')
    return infinite(num-1)

infinite(10)
#without using a while or for loop, lets you run it over and over again until you set a base case

10, looping again
9, looping again
8, looping again
7, looping again
6, looping again
5, looping again
4, looping again
3, looping again
2, looping again


'done'

#### Writing a Factorial Function

In [115]:
# 5! = 5*4*3*2*1 
def factorial(num):
    if num <=1:
        return num
    return num * factorial(num-1)
factorial(5)
#goes to find factorial(5) to 1 to find each value, then goes back up the line to multiply it

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [118]:
def subtraction(num):
    if num <=1:
        return num
    return num - subtraction(num-1)
subtraction(5)
#sub(5)= 5 - sub(4) which is 4- sub(3) which is 3 -sub(2) which is 1
#2-1 = 1; 3-1 = 2; 4-2 =2; 5-2 =3

3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [127]:
#for i in (range(10)):
#     print(i)

def myRange(start,stop,step=2):
    while start<stop:
        yield start  #yield keynote denotes a generator 
        start += step

#list(myRange(2,20))
myRange(2,20)
for i in myRange(2,20):
    myGeneratorValue = i
    print (i)

print(myGeneratorValue)


2
4
6
8
10
12
14
16
18
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [136]:
def newGenerator(start, stop, step=1):
    while start < stop:
        yield start
        start += step

for i in newGenerator(1,11):
    i = i**2
    myGeneratorValue= i
    print(i)
    
print(myGeneratorValue)

print('======================= think this is better')
def myGen(num):
    while num>0:
        yield num**2
        num -=1 
for x in myGen(10):
    print(x)
print(list(myGen(10)))

#why use a generator instead of just printing x**2 or appending?
#generators are good for working w/ huge data sets you don't want stored into a list; can convery to a list if you want
#everyday purposes- don't need to use generators, but if your code is slow, use generator


#need to really understand loops to get good at lambdas and maps


1
4
9
16
25
36
49
64
81
100
100
100
81
64
49
36
25
16
9
4
1
[100, 81, 64, 49, 36, 25, 16, 9, 4, 1]


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [1]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
 

new_places = [x.strip(' ') for x in places]
list(filter(None, new_places))


#empty items return false, but " "(spaces) aren't empty so must strip it before

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Andrew P. Garfield', 'David hassELHOFF', 'Shoha Tsuchida']`

In [22]:
author = ["Shoha Tsuchida", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
# new_author = author.title()

author = sorted(author, key=lambda x: x.lower().split()[-1])
print(author)

#key(optional) is a function to execute to decide the order; #reverse(optional): boolean w/ False sort ascending(default) and True sort descending
#sorted() accepts any iterable while list.sort() is only for lists
#syntax for sorted(): sorted(<iterable> <key> <reverse>)

        
    

['Victor aNisimov', 'Gary A.J. Bernstein', 'Andrew P. Garfield', 'David hassELHOFF', 'Shoha Tsuchida']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [38]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# class Farenheit:
#     def __init__(self, location, temp):
#         self.location = location
#         self.temp = temp

list(map(lambda x: (9/5) * x[1] + 32, places))

#still not very comfortable with maps


[89.6, 53.6, 111.2, 84.2]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [40]:
def fibNum(num):
    if num <=1:
        print(f'iteration{num}')
        return num
    print(f'Iteration{num}: ')
    return fibNum(num-1) + fibNum(num-2)
print(fibNum(5))


#need help for this as well

#0,1,1,2,3,5,8,13

Iteration5: 
Iteration4: 
Iteration3: 
Iteration2: 
iteration1
iteration0
iteration1
Iteration2: 
iteration1
iteration0
Iteration3: 
Iteration2: 
iteration1
iteration0
iteration1
5
